core/src/com/etheller/warsmash/viewer5/handlers/w3x/simulation/abilities/types/definitions/impl/CAbilityTypeDefinitionItemReincarnation.java

```diff
@@ -0,0 +1,38 @@
+package com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.types.definitions.impl;
+
+import com.etheller.warsmash.units.manager.MutableObjectData.MutableGameObject;
+import com.etheller.warsmash.util.War3ID;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.types.CAbilityType;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.types.definitions.CAbilityTypeDefinition;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.types.impl.CAbilityTypeItemReincarnation;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.abilities.types.impl.CAbilityTypeItemReincarnationLevelData;
+import com.etheller.warsmash.viewer5.handlers.w3x.simulation.combat.CTargetType;
+
+import java.util.EnumSet;
+import java.util.List;
+
+public class CAbilityTypeDefinitionItemReincarnation
+		extends AbstractCAbilityTypeDefinition<CAbilityTypeItemReincarnationLevelData> implements CAbilityTypeDefinition {
+	protected static final War3ID DELAY = War3ID.fromString("Ircd");
+	protected static final War3ID RESTORED_LIFE = War3ID.fromString("irc2");
+	protected static final War3ID RESTORED_MANA = War3ID.fromString("irc3");
+
+	@Override
+	protected CAbilityTypeItemReincarnationLevelData createLevelData(final MutableGameObject abilityEditorData,
+			final int level) {
+		final String targetsAllowedAtLevelString = abilityEditorData.getFieldAsString(TARGETS_ALLOWED, level);
+		final int delay = abilityEditorData.getFieldAsInteger(DELAY, level);
+		final int restoredLife = abilityEditorData.getFieldAsInteger(RESTORED_LIFE, level);
```

Comment:
In my other comment I suggested that you could delete the AbilityTypeDefinition and AbilityType for reincarnation if you were going to use the classes with "Spell" in the name, but now that I'm looking more closely I see this is for the Item version. You could probably have the Item version also be a subclass of `CAbilityPassiveSpellBase` if a base type like that were added. Then by implementing the `CAbilitySpell` interface, it would be able to use the slightly simpler API and skip the scaffolding, similar to what the hero skill did. There is not a requirement for the code to be one way or the other; it's currently all based on interfaces, and the decision whether to manually add `AbilityType` and `AbilityTypeDefinition` classes (or to just `implement CAbilitySpell` ) is basically a matter of whether the implementer wants a possible slight performance gain from a cached symbol table (i.e. POJO class) for the "ability type" level data.

Reply:
But in practice, it's not split on what _should_ be more performant, but rather based on the fact that newer classes implement `CAbilitySpell` so that I could write them more quickly and get more done.